crates/ironrdp-graphics/src/color_conversion.rs

```diff
@@ -1,153 +1,43 @@
-
-fn xrgb_to_rgb(pixel: &[u8]) -> Rgb {
-    Rgb {
-        r: pixel[1],
-        g: pixel[2],
-        b: pixel[3],
-    }
-}
-
-fn xbgr_to_rgb(pixel: &[u8]) -> Rgb {
-    Rgb {
-        b: pixel[1],
-        g: pixel[2],
-        r: pixel[3],
-    }
-}
-
-fn bgrx_to_rgb(pixel: &[u8]) -> Rgb {
-    Rgb {
-        b: pixel[0],
-        g: pixel[1],
-        r: pixel[2],
-    }
-}
-
-fn rgbx_to_rgb(pixel: &[u8]) -> Rgb {
-    Rgb {
-        r: pixel[0],
-        g: pixel[1],
-        b: pixel[2],
-    }
-}
-
-const fn pixel_format_to_rgb_fn(format: PixelFormat) -> fn(&[u8]) -> Rgb {
-    match format {
-        PixelFormat::ARgb32 | PixelFormat::XRgb32 => xrgb_to_rgb,
-        PixelFormat::ABgr32 | PixelFormat::XBgr32 => xbgr_to_rgb,
-        PixelFormat::BgrA32 | PixelFormat::BgrX32 => bgrx_to_rgb,
-        PixelFormat::RgbA32 | PixelFormat::RgbX32 => rgbx_to_rgb,
-    }
-}
-
-#[allow(clippy::too_many_arguments)]
-pub fn to_ycbcr(
-    mut input: &[u8],
-    width: usize,
-    height: usize,
-    stride: usize,
-    format: PixelFormat,
-    mut y: &mut [i16],
-    mut cb: &mut [i16],
-    mut cr: &mut [i16],
-) {
-    let to_rgb = pixel_format_to_rgb_fn(format);
-    let bpp = format.bytes_per_pixel() as usize;
-
-    for _ in 0..height {
-        iter_to_ycbcr(input[..width * bpp].chunks_exact(bpp), y, cb, cr, to_rgb);
-        input = &input[stride..];
-        y = &mut y[width..];
-        cb = &mut cb[width..];
-        cr = &mut cr[width..];
-    }
-}
-
-struct TileIterator<'a> {
-    slice: &'a [u8],
-    x: usize,
-    y: usize,
-    width: usize,
-    height: usize,
-    stride: usize,
-    bpp: usize,
-}
-
-impl<'a> TileIterator<'a> {
-    fn new(slice: &'a [u8], width: usize, height: usize, stride: usize, bpp: usize) -> Self {
-        assert!(width >= 1);
-        assert!(height >= 1);
-
-        Self {
-            slice,
-            x: 0,
-            y: 0,
-            width,
-            height,
-            stride,
-            bpp,
-        }
-    }
-}
-
-impl<'a> Iterator for TileIterator<'a> {
-    type Item = &'a [u8];
-
-    fn next(&mut self) -> Option<Self::Item> {
-        // create 64x64 tiles
-        if self.y >= 64 {
-            return None;
-        }
+use yuvutils_rs::{
+    rdp_abgr_to_yuv444, rdp_argb_to_yuv444, rdp_bgra_to_yuv444, rdp_rgba_to_yuv444, rdp_yuv444_to_argb,
+    rdp_yuv444_to_rgba, BufferStoreMut, YuvPlanarImage, YuvPlanarImageMut,
+};
 
-        // repeat the last column & line if necessary
-        let y = min(self.y, self.height - 1);
-        let x = min(self.x, self.width - 1);
-        let pos = y * self.stride + x * self.bpp;
-
-        self.x += 1;
-        if self.x >= 64 {
-            self.x = 0;
-            self.y += 1;
-        }
-
-        Some(&self.slice[pos..pos + self.bpp])
-    }
-}
+use crate::image_processing::PixelFormat;
 
-impl ExactSizeIterator for TileIterator<'_> {
-    fn len(&self) -> usize {
-        64 * 64
-    }
+// used for the test suite, we may want to drop it
```

Comment:
suggestion: Add a FIXME prefix so it’s easier to remember about this.

Reply:
ok